File: lexer/src/main/java/net/jr/lexer/impl/QuotedString.java
package net.jr.lexer.impl;

import static net.jr.lexer.impl.CharConstraint.Builder.*;

public class QuotedString extends LexemeImpl {

    private Automaton automaton;

    private char starChar;

    private char endChar;

    private char escapeChar;

    private char[] forbiddenChars;

    public QuotedString(char startChar, char endChar, char escapeChar, char[] forbiddenChars) {
        this.starChar = startChar;
        this.endChar = endChar;
        this.escapeChar = escapeChar;
        this.forbiddenChars = forbiddenChars;
        DefaultAutomaton.Builder builder = DefaultAutomaton.Builder.forTokenType(this);
        DefaultAutomaton.Builder.BuilderState inString = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState escaping = builder.newNonFinalState();
        builder.initialState().when(eq(startChar)).goTo(inString);
        inString.when(eq(escapeChar)).goTo(escaping);
        inString.when(inList(forbiddenChars)).goTo(builder.failedState());
        inString.when(not(eq(endChar))).goTo(inString);
        escaping.when(any()).goTo(inString);
        inString.when(eq(endChar)).goTo(builder.newFinalState());
        this.automaton = builder.build();
    }

    @Override
    public int getPriority() {
        return 1;
    }

    @Override
    public Automaton getAutomaton() {
        return automaton;
    }
}

File: src/lexer/QuotedString.ts
import { Terminal } from "../common/Terminal";
import { Automaton, AutomatonBuilder } from "./automaton/Automaton";
import { CharConstraint } from "./CharConstraint";

export class QuotedString extends Terminal {

	private _startChar: string;

	private _endChar: string;

	private _escapeChar: string;

	private _forbiddenChars: string;

	private _automaton: Automaton;

	constructor(startChar: string, endChar: string, escapeChar: string, forbiddenChars: string) {
		super("QuotedString");
		this._startChar = startChar;
		this._endChar = endChar;
		this._escapeChar = escapeChar;
		this._forbiddenChars = forbiddenChars;


		let builder = AutomatonBuilder.forTokenType(this);
		let start = builder.initialState();

		//must begin with the 'start' char
		let inString = builder.newNonFinalState();
		builder.addTransition(start, CharConstraint.eq(startChar), inString);

		let end = builder.newFinalState();

		//get out if we encounter the 'end' char
		builder.addTransition(inString, CharConstraint.not(CharConstraint.inList(forbiddenChars + endChar + escapeChar)), inString);
		builder.addTransition(inString, CharConstraint.eq(endChar), end);

		//escaping
		let escaping = builder.newNonFinalState();

		let constraint = forbiddenChars.length ? CharConstraint.not(CharConstraint.inList(forbiddenChars)) : CharConstraint.any();
		builder.addTransition(inString, CharConstraint.eq(escapeChar), escaping);
		builder.addTransition(escaping, constraint, inString);

		// otherwise continue reading the string
		//inString.addFallback().target = inString;

		this._automaton = builder.build();
	}

	get automaton(): Automaton {
		return this._automaton;
	}
};